ci: Run Windows native task on GitHub Actions

hebasto opened the pull request:
From https://github.com/bitcoin/bitcoin/issues/28098:
> Thus, someone would have to sponsor an amount of roughly 5kUSD/mo for those two tasks.

> If the goal is to stay on a free plan, I think the only option is GitHub Actions CI.

Historical context:
- https://github.com/bitcoin/bitcoin/pull/17697
- https://github.com/bitcoin/bitcoin/issues/17803
- https://github.com/bitcoin/bitcoin/pull/18031

Security concerns:
- https://github.com/bitcoin/bitcoin/issues/28098#issuecomment-1651432106
- https://github.com/bitcoin/bitcoin/issues/28098#issuecomment-1651688197

`GITHUB_TOKEN` permissions (from the build log in my personal repo):
```
2023-07-27T07:30:17.8313534Z ##[group]GITHUB_TOKEN Permissions
2023-07-27T07:30:17.8314113Z Contents: read
2023-07-27T07:30:17.8314608Z Metadata: read
2023-07-27T07:30:17.8314957Z Packages: read
2023-07-27T07:30:17.8315233Z ##[endgroup]
```

Comparison of resources:

| Resource | Current, Cirrus CI | Suggested, GitHub Actions |
|---|:-:|:-:|
| CPU | 6 | 2 |
| RAM, GB | 12 | 7 |

The `TEST_RUNNER_TIMEOUT_FACTOR` variable is set to the current default value for all CI tasks: https://github.com/bitcoin/bitcoin/blob/64440bb733896a7a2caf902825e0406cb993e666/ci/test/00_setup_env.sh#L48

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,190 @@
+
+name: bitcoin-core-ci
+run-name: Bitcoin Core CI
+on: [pull_request, push]
```
Done.

hebasto commented:
> Can you remove the Windows Cirrus task here, as it will need to be removed either way?

Done.

> Maybe also print the `GITHUB_TOKEN` permissions at every start of the task, just to double check they are read-only?

They are printed by default. To see them, in https://github.com/hebasto/bitcoin/actions/runs/5685393993/job/15410187833, one needs to expand "Set up job", then expand "GIHUB_TOKEN Permissions".

hebasto commented:
> In the meantime it may be good to do 10 runs and then check how many of them fail

Observing intermittent "Error: no RPC connection". Converting this PR to a draft for now

hebasto commented:
> In the meantime it may be good to do 10 runs and then check how many of them fail

10 runs are green for the recent commit:
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/1
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/2
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/3
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/4
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/5
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/6
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/7
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/8
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/9
- https://github.com/hebasto/bitcoin/actions/runs/5701707622/attempts/10

> Ok, so I guess someone needs to enable Actions in both orgs for this to proceed.

@sipa @fanquake 

Mind enabling GitHub Actions in the https://github.com/bitcoin/bitcoin repository?

I'm suggesting the repository "Actions permissions" as follows:

![image](https://github.com/bitcoin/bitcoin/assets/32963518/c4ff04f9-2991-4c23-b268-8cffcde03a74)

![image](https://github.com/bitcoin/bitcoin/assets/32963518/5fe273c2-59d5-4517-b3f5-0ce2ee9f44fd)

hebasto commented:
Converting this PR to a draft as it conflicts with https://github.com/bitcoin/bitcoin/pull/28187.

Please review https://github.com/bitcoin/bitcoin/pull/28187 first.

hebasto commented:
https://github.com/bitcoin/bitcoin/pull/28292 is a blocker for this PR.

hebasto commented:
> > #28292 is a blocker for this PR.
> 
> Is it? It should only affect the cache hit percentage, no?

Yes, that is what I mean. Without #28292, caches for the master branch are to be evicted very often. And this PR makes it roughly 2 times more often.

hebasto commented:
Rebased.

hebasto commented:
Rebased.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -51,12 +51,206 @@ jobs:
+      - name: Run rpcauth test
+        run: python test\util\rpcauth-test.py
+
+      - name: Run functional tests
```
Now it looks like that:

![image](https://github.com/bitcoin/bitcoin/assets/32963518/7eefbf99-1f31-41f7-ad16-254dcc567a47)
 
and

![image](https://github.com/bitcoin/bitcoin/assets/32963518/3b5dfbb2-a616-411e-ba38-c0d465dfb6fb)

So the command itself is available in the logs.

Besides, some commands are pretty long, for instance:
```
python test\functional\test_runner.py --ci --tmpdirprefix=%RUNNER_TEMP% --combinedlogslen=99999999 --timeout-factor=%TEST_RUNNER_TIMEOUT_FACTOR% --quiet
```

> Any reason to give each step a name?

I think, it's a matter of consistency and personal taste.

In any case, I'm happy to update this PR and remove the `name` property from specific steps upon your request.

maflcko reviewed (commented):
Looks like CI doesn't run on this? I wonder if it makes sense to document the required permissions inside the file?

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -51,12 +51,206 @@ jobs:
+        env:
+          PYTHONUTF8: 1
+        shell: cmd
+        run: python test\functional\test_runner.py --ci --tmpdirprefix=%RUNNER_TEMP% --combinedlogslen=99999999 --timeout-factor=%TEST_RUNNER_TIMEOUT_FACTOR% --quiet
```
Well, a couple of tests run [too long](https://github.com/hebasto/bitcoin/actions/runs/5927065845/job/16069764131):
```
feature_index_prune.py                                 | ✓ Passed  | 3307 s
feature_pruning.py                                     | ✓ Passed  | 2845 s
```

hebasto replied:
https://cirrus-ci.com/task/5095403331256320:
```
feature_index_prune.py                                 | ✓ Passed  | 2695 s
feature_pruning.py                                     | ✓ Passed  | 2409 s
```

hebasto replied:
> An alternative would be to run them on non-pr pushes only. Failures should be rare enough to deal with them post-merge.

Done in https://github.com/bitcoin/bitcoin/pull/28360.

real-or-random commented:
> I wonder if it makes sense to document the required permissions inside the file?

I voted against documenting them [here](https://github.com/bitcoin-core/secp256k1/pull/1408#issuecomment-1687760351).

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -51,12 +51,213 @@ jobs:
+      - name: Run benchmarks
+        run: src\bench_bitcoin.exe -sanity-check
+
+      - name: Run bitcoin utils tests
```
Done.

hebasto commented:
> Looks like CI doesn't run on this?

@fanquake 

I'm kindly asking you to add [`ilammy/msvc-dev-cmd@*`](https://github.com/marketplace/actions/enable-developer-command-prompt) to the repository's Actions permissions.

> > I wonder if it makes sense to document the required permissions inside the file?
> 
> I voted against documenting them [here](https://github.com/bitcoin-core/secp256k1/pull/1408#issuecomment-1687760351).

Your opinion about this will be very appreciated as well.

hebasto commented:
> Looks like CI doesn't run on this?

It runs now: https://github.com/bitcoin/bitcoin/actions/runs/5949997773/job/16136915568

@fanquake Thank you so much!

hebasto commented:
> Please squash your commits according to [`master`/CONTRIBUTING.md#squashing-commits](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md?rgh-link-date=2023-08-23T12%3A07%3A33Z#squashing-commits)

Squashed your commits according to [`master`/CONTRIBUTING.md#squashing-commits](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md?rgh-link-date=2023-08-23T12%3A07%3A33Z#squashing-commits) :D

dergoegge reviewed (approved):
lgtm ACK b0e233ba3b61f2ea40b6a5a7534f49a0d4ad4568

dergoegge commented:
> By the way, what's an "lgtm ACK"?

(at least for me) it's just a normal ack pre-fixed with "looks good to me"

dergoegge reviewed (approved):
reACK a3ebaa20eaab37b9ec32d6569fbc63d6216fe1b5

hebasto commented:
Friendly ping @MarcoFalke :)

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -68,3 +68,200 @@ jobs:
+          key: ${{ github.job }}-vcpkg-tools
+
+      - name: vcpkg binary cache
+        uses: actions/cache@v3
```
While `key` remains the same, pull requests will restore cache from the master branch and make no new saves.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -68,3 +68,200 @@ jobs:
+
+      - name: Run functional tests
+        env:
+          PYTHONUTF8: 1
```
> why only set this for the fun tests?

Because only functional tests require it.

> This should be global

Which python scripts for?

hebasto replied:
> The same level that was used in Cirrus. No need to change anything here.

Done.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -68,3 +68,200 @@ jobs:
+        with:
+          path: ~/AppData/Local/ccache
+          key: ${{ github.job }}-ccache-${{ github.run_id }}
+          restore-keys: ${{ github.job }}-ccache-
```
No, it won't.

It is the "Restore" action. The paired "Save" action is guarded with https://github.com/bitcoin/bitcoin/blob/a3ebaa20eaab37b9ec32d6569fbc63d6216fe1b5/.github/workflows/ci.yml#L245

hebasto committed:
ci: Run "Win64 native" job on GitHub Actions

hebasto commented:
Updated a3ebaa20eaab37b9ec32d6569fbc63d6216fe1b5 -> 6a4337298035683a8748ae446eae90e5c5f41d1b ([pr28173.11](https://github.com/hebasto/bitcoin/commits/pr28173.11) -> [pr28173.12](https://github.com/hebasto/bitcoin/commits/pr28173.12)):

- rebased
- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28173#discussion_r1305602588)

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -68,3 +68,199 @@ jobs:
+        uses: actions/cache@v3
+        with:
+          path: C:/vcpkg/downloads/tools
+          key: ${{ github.job }}-vcpkg-tools
```
https://github.com/bitcoin/bitcoin/pull/28173#discussion_r1307907257

hebasto commented on .cirrus.yml:
```diff
@@ -101,102 +101,6 @@ task:
-    # Increase the dynamic port range to the maximum allowed value to mitigate "OSError: [WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted".
-    # See: https://learn.microsoft.com/en-us/biztalk/technical-guides/settings-that-can-be-modified-to-improve-network-performance
-    - netsh int ipv4 set dynamicport tcp start=1025 num=64511
-    - netsh int ipv6 set dynamicport tcp start=1025 num=64511
```
No, logs say nothing about `WinError 10048` error.

See https://github.com/bitcoin/bitcoin/pull/28384 for a proposed fix.
